### PR TITLE
[SYCL][Bindless] improve vulkan interop unsampled images testing

### PR DESCRIPTION
**Pull request by cppchedy**
Improve the testing of Vulkan interop unsampled images by extending the test to run with different combination of parameters such as image size, number of dimensions, etc..

### CODE REVIEW

File: sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp [hunk 1]
```diff
@@ -14,246 +14,468 @@
...
+    switch (order) {
+    case sycl::image_channel_order::r:
+      return VK_FORMAT_R32_SINT;
+    case sycl::image_channel_order::rg:
+      return VK_FORMAT_R32G32_SINT;
+    case sycl::image_channel_order::rgba:
+      return VK_FORMAT_R32G32B32A32_SINT;
+    default: {
+      std::cerr << "error in converting to vulkan format"
+                << "\n";
```
[the remark this answers is not in the document]
Comment (aelovikov-intel) [reply sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp:50]: Same in other places.

File: sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp [hunk 2]
```diff
@@ -14,246 +14,468 @@
...
+  distribution_t distribution(static_cast<DType>(0), static_cast<DType>(100));
+
+  for (int i = 0; i < v.size(); ++i) {
+    v[i] = sycl::vec<DType, NChannels>(distribution(generator));
+  }
+}
+
+template <typename DType, int NChannels>
+std::conditional_t<NChannels == 1, DType, sycl::vec<DType, NChannels>>
+add_kernel(std::conditional_t<NChannels == 1, DType,
```
Comment (aelovikov-intel) [inline sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp:193]: I'd do

```
T add_kernel(T) {
  if constexpr (std::is_scalar_v<T>)
```

or something similar.

File: sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp [hunk 3]
```diff
@@ -14,246 +14,468 @@
...
-                sycl::ext::oneapi::experimental::read_image<sycl::uint4>(
-                    img_input, sycl::int2(dim0, dim1));
-
-            pixel *= 10;
-
-            // Extension: write image data using handle (Vulkan imported)
-            sycl::ext::oneapi::experimental::write_image(
-                img_output, sycl::int2(dim0, dim1), pixel);
+            if constexpr (NDims == 2) {
+
```
Comment (aelovikov-intel) [inline sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp:254]:
```suggestion
```

File: sycl/test-e2e/bindless_images/vulkan_interop/sampled_images.cpp [hunk 4]
```diff
@@ -184,16 +184,17 @@ bool run_test() {
...
 
-    VK_CHECK_CALL(vkBeginCommandBuffer(vk_computeCmdBuffer, &cbbi));
-    vkCmdPipelineBarrier(vk_computeCmdBuffer, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,
+    VK_CHECK_CALL(vkBeginCommandBuffer(vk_compute_cmd_buffer, &cbbi));
+    vkCmdPipelineBarrier(vk_compute_cmd_buffer,
+                         VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,
                          VK_PIPELINE_STAGE_TRANSFER_BIT, 0, 0, nullptr, 0,
                          nullptr, 1, &barrierInput);
-    VK_CHECK_CALL(vkEndCommandBuffer(vk_computeCmdBuffer));
+    VK_CHECK_CALL(vkEndCommandBuffer(vk_compute_cmd_buffer));
```
Comment (aelovikov-intel) [inline sycl/test-e2e/bindless_images/vulkan_interop/sampled_images.cpp:192]: Merely changing code style isn't "improve vulkan interop ... testing" and should be done in a separate NFC PR.

File: sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp [hunk 5]
```diff
@@ -14,246 +14,468 @@
...
+  sycl::ext::oneapi::experimental::unsampled_image_handle input_1, input_2,
+      output;
+};
+
+handles_t
+create_test_handles(sycl::context &ctxt, sycl::device &dev,
+                    int input_image_fd_1, int input_image_fd_2,
+                    int output_image_fd, int sycl_wait_semaphore_fd,
+                    int sycl_done_semaphore_fd, const size_t img_size,
+                    sycl::ext::oneapi::experimental::image_descriptor &desc) {
```
[the remark this answers is not in the document]
Comment (cppchedy) [reply sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp:97]: hey @aelovikov-intel, For `handles_t` struct I need to store sycl context and device, which will make it not device copyable. 
I removed the helpers function for `vulkan_image_test_resources_t` struct, which not used in the kernel, and integrated everything as you suggested.

File: sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (cppchedy) [reply sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp]: applied.

File: sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp [hunk 6]
```diff
@@ -14,246 +14,468 @@
...
+  std::default_random_engine generator;
+  using distribution_t =
+      std::conditional_t<std::is_integral_v<DType>,
+                         std::uniform_int_distribution<DType>,
+                         std::uniform_real_distribution<DType>>;
+  distribution_t distribution(static_cast<DType>(0), static_cast<DType>(100));
+
+  for (int i = 0; i < v.size(); ++i) {
+    v[i] = sycl::vec<DType, NChannels>(distribution(generator));
+  }
```
[the remark this answers is not in the document]
Comment (cppchedy) [reply sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp]: applied.

File: sycl/test-e2e/bindless_images/vulkan_interop/sampled_images.cpp (hunk 4 above)
Comment (cppchedy) [reply sycl/test-e2e/bindless_images/vulkan_interop/sampled_images.cpp]: reverted to include only functional changes. other changes will be part of a separate PR as you suggested.

File: sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp [hunk 7]
```diff
@@ -433,12 +705,12 @@ int main() {
...
-  bool validated = run_test();
+  auto exit_value = run_all();
 
   if (vkutil::cleanup() != VK_SUCCESS) {
     std::cerr << "Cleanup failed!\n";
     return EXIT_FAILURE;
   }
 
-  return validated ? EXIT_SUCCESS : EXIT_FAILURE;
+  return !exit_value;
```
Comment (aelovikov-intel) [inline sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp:715]: This is very strange... Can we rename the variable?

**Review by aelovikov-intel — APPROVED** [review]
I'll leave the detailed review to the component owner (@bindless-images-reviewers).

File: sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp (hunk 7 above)
Comment (cppchedy) [reply sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp]: changed to `run_ok`.

File: sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp [hunk 8]
```diff
@@ -14,228 +14,429 @@
...
+  if (NDims > 1) {
+    num_elems *= dims[1];
+    height = static_cast<uint32_t>(dims[1]);
+    imgType = VK_IMAGE_TYPE_2D;
+  }
+  if (NDims > 2) {
+    num_elems *= dims[2];
+    depth = static_cast<uint32_t>(dims[2]);
+    imgType = VK_IMAGE_TYPE_3D;
+  }
```
Comment (przemektmalon) [inline sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp:388]:
```suggestion
  if constexpr (NDims == 1) {
    num_elems *= dims[1];
    height = static_cast<uint32_t>(dims[1]);
    imgType = VK_IMAGE_TYPE_2D;
  } else if constexpr (NDims == 2) {
    num_elems *= dims[2];
    depth = static_cast<uint32_t>(dims[2]);
    imgType = VK_IMAGE_TYPE_3D;
  }
```

**Review by przemektmalon — APPROVED** [review]

File: sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp [hunk 9]
```diff
@@ -370,51 +583,110 @@ bool run_test() {
...
-  for (int i = 0; i < width * height; ++i) {
-    sycl::vec<uint32_t, 4> expected = {4 * i, 4 * i + 1, 4 * i + 2, 4 * i + 3};
-    expected *= 10;
-    for (int j = 0; j < 4; ++j) {
-      if (outputStagingData[i][j] != expected[j]) {
-#ifdef VERBOSE_PRINT
-        std::cout << "Result mismatch! actual[" << i << "][" << j
+  for (int i = 0; i < num_elems; ++i) {
+    VecType expected = input_vector_0[i];
+    expected += input_vector_1[i];
```
[the remark this answers is not in the document]
Comment (cppchedy) [reply sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp]: applied.

**Comment by ProGTX** [comment]
@intel/llvm-gatekeepers ready to merge

**Pull request merged**